helm-v3/006-repositories.md

```diff
@@ -1,40 +1,177 @@
+
+### `/.well-known/helm`
+
+The Capabilities API will be enabled for a specific domain via
+[Well-Known URI](https://tools.ietf.org/html/rfc5785) located at
+`/.well-known/helm`.
+
+*Note: This file may also serve to provide configuration for other APIs
+in the future by supplying an additional top-level section.*
+
+The response body should be valid JSON and should supply a valid,
+top-level `capabilities` section. For example:
+
+```
+{
+    "capabilities": {
+        "login": {
+            ...
+        },
+        "push": {
+            ...
+        },
+        "search": {
+            ...
+        }
+    }
+}
+```
+
+Each subsection (i.e. `login`, `push`, `search`) will provide configuration
+specific to the capabiltity.
+
+For example, the following might inform Helm that this domain only
+accepts token-based login:
+
+```
+...
+        "login": {
+            "methods": {
+                "token": {
+                    "loginUrl": "https://auth.site.com/oauth2/authorize",
+                    "tokenUrl": "https://auth.site.com/oauth2/token"
+                }
+            }
+        }
+...
+```
+
+## The `helm login` command
+
+A new `helm login` command will be added to Helm, responsible for
+managing authentication across repository providers.
+
+The basic usage will be:
+
+```
+helm login <url>
+```
+
+This command is intended to allow users to authenticate just one time in
+order to access multiple repositories provided by the same source. 
+
+Afterwards, running a `helm repo add` pointing to any repository URL
+tied to a provider that has been successfully authenticated against will
+not require additional credentials.
+
+The exact semantics are TBD, but several auth methods will be supported,
+such as:
+
+* Basic Auth
+* Cert based Auth
+* Token/Bearer Auth
+
+### Storing Credentials
+
+Credentials obtained via `helm login` will be stored in a new `providers`
+section in `$HELM_HOME/repository/repositories.yaml`.
+
+For all provider entries, the following field is required:
+* `domain` - the FQDN of the original URL specified during `helm login`
+
+Other fields will be stored based on the authentication method used.
+
+The following example shows a snippet of  what `repositories.yaml` might
+look like with a `providers` section:
+```
+apiVersion: v1
+generated: 2018-10-25T17:02:41.478448894-05:00
+providers:
+- domain: site.com
+  realm: https://site.com/oauth2/helm
+  expires: 2018-10-26T17:02:41.478448894-05:00
+  accessToken: MTQ0NjJkZmQ5OTM2NDE1ZTZjNGZmZjI3
+  refreshToken: IwOGYzYTlmM2YxOTQ5MGE3YmNmMDFkNTVk
+repositories:
+...
+```
+
+## The `helm push` command
+
+A new `helm push` command will be added to Helm, responsible for uploading
+chart versions to a repository.
+
+The basic usage will be:
+
+```
+helm push <chart> <repo>
+```
+
+The exact semantics of this command are still TBD.
+
+
+The most basic push method supported is referred to as `http`, and should be
```

Comment:
is this being explicit that https is a different configuration?

Reply:
According to the below section, yes, https could be a different configuration than http. You could even support only one of the two protocols (such as if you don't want to support the http protocol, but want to support https).

Reply:
note that this matches with the current getter plugin model in Helm. Plugins can register for different getter/pusher methods. For example `helm push cm://mychartmuseum.com/foo` uses the `cm` protocol, which then the repository can choose to accept (as in, it has configuration available for the `cm` method), or deny (missing from `/.well-known/helm` endpoint)

Reply:
Yes. I did not add anything about plugins here to limit the scope, but I would imagine plugins could also use Capabilities API to check for compatibility and obtain configuration.